File: examples/lmm_penicillin.jl
# Please install the required packages before running:
# import Pkg; Pkg.add(["CSV", "DataFrames", "MixedModels"])

using CSV
using DataFrames
using MixedModels

function main()
    file_path = joinpath("tests", "data", "penicillin.csv")
    if !isfile(file_path)
        println("Could not find the dataset at: $file_path")
        println("Please run this example from the root of the lme-rs repository.")
        exit(1)
    end

    println("Loading data from $file_path...")
    
    # 1. Load the dataset
    df = CSV.read(file_path, DataFrame)

    # 2. Fit the Linear Mixed Model
    println("\nFitting model: diameter ~ 1 + (1 | plate) + (1 | sample)")
    
    form = @formula(diameter ~ 1 + (1 | plate) + (1 | sample))
    
    # Fit the Linear Mixed Model (REML = true by default in fit)
    m1 = fit(MixedModel, form, df, REML=true)
    
    # 3. Print the summary
    println("\n=== Model Summary ===")
    println(m1)

    # 4. Generate Predictions
    println("\n=== Predictions ===")
    println("Generating predictions for new plates and samples...")
    
    newdata = DataFrame(
        plate = ["a", "b", "c", "d"],
        sample = ["A", "C", "E", "F"]
    )
    
    # Predict (Population-level)
    # \mu = X \beta
    beta = coef(m1)
    X = ones(4, 1) # Intercept only for fixed effects
    
    preds = X * beta
    
    println("Predictions (Population-level):")
    println(preds)
end

main()

File: examples/lmm_penicillin.rs
use lme_rs::lmer;
use polars::prelude::*;

fn main() -> anyhow::Result<()> {
    // 1. Load the dataset
    let mut file = std::fs::File::open("tests/data/penicillin.csv")?;
    let df = CsvReadOptions::default()
        .with_has_header(true)
        .into_reader_with_file_handle(&mut file)
        .finish()?;

    // Cast string columns
    let df = df
        .lazy()
        .with_columns(vec![
            col("plate").cast(DataType::String),
            col("sample").cast(DataType::String),
        ])
        .collect()?;

    println!("Fitting model: diameter ~ 1 + (1 | plate) + (1 | sample)");

    // 2. Fit the model
    let formula = "diameter ~ 1 + (1 | plate) + (1 | sample)";
    let fit = lmer(formula, &df, true)?;

    // 3. Print the summary
    println!("\n=== Model Summary ===");
    println!("{}", fit);

    println!("\n=== Predictions ===");
    println!("Generating predictions for new plates and samples...");

    // 4. Generate Predictions
    let new_plate = Series::new("plate".into(), &["a", "b", "c", "d"]);
    let new_sample = Series::new("sample".into(), &["A", "C", "E", "F"]);

    let newdata = DataFrame::new(vec![new_plate.into(), new_sample.into()])?;

    let preds = fit.predict(&newdata)?;
    println!("Predictions:\n{:?}", preds);

    Ok(())
}
